extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,389 @@
+
+    if (data['showNotification']) {
+      return Promise.resolve(data['showNotification']);
+    }
+    // if no notification is shown, resolve right away with `false` since
+    // we dont need to show it.
+    return this.dialogResolve_(false);
+  }
+
+  /**
+   * Get async cid service.
+   * @return {!Promise}
+   * @private
+   */
+  getAsyncCid_() {
+    this.dialogPromise_ = new Promise((resolve, reject) => {
+      this.dialogResolve_ = resolve;
+    });
+    return cidFor(this.win_).then(cid => {
+      // `amp-user-notification` is our cid scope, while we give it a resolved
+      // promise for the 2nd argument so that the 3rd argument (the
+      // persistentConsent) is the one used to resolve getting
+      // the external CID.
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getAsyncCid_()
+        .then(this.postShowEndpoint_.bind(this))
+        .then(this.onPostShowEndpointSuccess_.bind(this));
+  }
+
+  /**
+   * Shows the current user notification. Adds the `.amp-active` class to
+   * the element.
+   */
+  show() {
+    this.element.style.display = 'block';
+    this.element.classList.add('amp-active');
+    return this.dialogPromise_;
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.style.display = 'none';
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_(false);
+    this.postDismissEnpoint_();
+  }
+}
+
+
+/**
+ * UserNotificationManager handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationManager {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDeferDef>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.documentReadyPromise_ = new Promise((resolve) => {
+      this.documentReadyResolve_ = resolve;
+    });
+
+    // TODO: Add `whenDocumentReady` function in document-state
+    // that returns a promise.
+    // See https://github.com/ampproject/amphtml/issues/1226
+    onDocumentReady(this.win_, () => {
+      this.documentReadyResolve_();
+    });
+
+    /** @private {!Promise} */
+    this.nextInQueue_ = this.viewer_.whenVisible().then(() => {
+      // we block on both viewer visible and when dom ready.
+      return this.documentReadyPromise_;
+    }).then(() => {
+      Object.keys(this.deferRegistry_).forEach((id) => {
+        assert(this.win_.document.getElementById(id) != null,
```

Comment:
I think we should instead go with warnings. One notification failing all the rest is pretty drastic.